File: algos/range_queries/PrefixSum/soln/Apoorv012/Solution3.cpp
/*

Ashish has a binary string s of length n that he wants to sort in non-decreasing order.
Choose a subsequence of any length such that its elements are in non-increasing order.
Reverse this subsequence in-place.

Time: O(n)
Space: O(n)

Link: https://codeforces.com/contest/1605/submission/355290481

*/

#include <bits/stdc++.h>
using namespace std;

using ll = long long;
const ll MOD = 998244353;

void solve() {
    int n;
    cin >> n;
    string s;
    cin >> s;

    // cout << "n, s: " << n << " " << s << endl;

    int st = 0, en = n-1;
    vector<int> ones, zeroes;
    while (st < en) {
        // cout << st << " -=- " << en << endl;
        bool oneFound = false, zeroFound = false;
        while (st < n && s[st] == '0') {
            st++;
        }
        if (st < n) oneFound = true;
        // cout << "found one at " << st << endl;
        while (en >= 0 && s[en] == '1') {
            en--;
        }
        if (en >= 0) zeroFound = true;
        // cout << "found zero at " << en << endl;
        // cout << oneFound << " " << zeroFound << endl;
        if (oneFound && zeroFound) {
            // cout << "meow\n";
            if (st < en) {
                // cout << "yes\n";
                ones.push_back(st++ + 1);
                zeroes.push_back(en-- + 1);
                continue;
            }
            else break;
        }
        else {
            break;
        }
    }

    if (!ones.size()) {
        cout << 0 << endl;
        return;
    }

    reverse(zeroes.begin(), zeroes.end());

    cout << 1 << endl;
    cout << (2 * ones.size()) << " ";
    for (auto el: ones) cout << el << " ";
    for (auto el: zeroes) cout << el << " ";

    cout << endl;

}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int t = 1;
    cin >> t;
    
    while (t--) {
        solve();
    }
    
    return 0;
}

File: CodeCraft-main/algos/range_queries/PrefixSum/soln/dwivediprashant/Solution3.java
/*
------------------------------------------------------------
 Problem: Reverse Sort
 Platform: Codeforces
 Contest: Codeforces Round #753 (Div. 3)
 Problem Link:
 https://codeforces.com/contest/1605/problem/B

 Submission Link:
 https://codeforces.com/contest/1605/submission/355299729
------------------------------------------------------------

 Problem Summary:
 ----------------
 Ashish has a binary string s of length n.
 He wants to sort it in non-decreasing order (all 0s first, then 1s).

 Allowed Operation:
 ------------------
 - Choose a subsequence (not necessarily contiguous).
 - The chosen subsequence must be in non-increasing order.
   (i.e., sequence of 1s followed by 0s is allowed)
 - Reverse this subsequence in-place.

 Goal:
 -----
 Find the minimum number of operations required to sort the string.
 It is guaranteed the answer is ≤ n.

 Key Insight:
 ------------
 - If the string is already sorted, no operation is needed.
 - Otherwise, the string can always be sorted in exactly ONE operation.
 - Construct the target sorted string:
      first all '0's, then all '1's
 - Collect all indices where the original string differs from the target.
 - These mismatched positions form a valid non-increasing subsequence:
      1s (that should be 0) followed by
      0s (that should be 1)
 - Reversing this subsequence fixes all mismatches in one operation.

 Time Complexity:
 ----------------
 O(n) per test case

 Space Complexity:
 -----------------
 O(n)

------------------------------------------------------------
*/

import java.io.*;
import java.util.*;

public class Solution3 {
    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder output = new StringBuilder();

        int t = Integer.parseInt(br.readLine().trim());

        while (t-- > 0) {

            int n = Integer.parseInt(br.readLine().trim());
            String s = br.readLine().trim();

            // Count number of zeros
            int zeroCount = 0;
            for (int i = 0; i < n; i++) {
                if (s.charAt(i) == '0') zeroCount++;
            }

            // Collect mismatched indices
            List<Integer> mismatch = new ArrayList<>();

            for (int i = 0; i < n; i++) {
                char expected = (i < zeroCount) ? '0' : '1';
                if (s.charAt(i) != expected) {
                    mismatch.add(i + 1); // 1-based indexing
                }
            }

            // If already sorted
            if (mismatch.isEmpty()) {
                output.append(0).append('\n');
            } else {
                // One operation is enough
                output.append(1).append('\n');
                output.append(mismatch.size());
                for (int idx : mismatch) {
                    output.append(" ").append(idx);
                }
                output.append('\n'); 
            }
        }

        System.out.print(output.toString());
    }
}
